File: 1kyu/simple_interactive_interpreter.js
class Token {
  constructor(expr) {
    const reg = /\s*([-+*\/%=\(\)]|[A-Za-z_][A-Za-z0-9_]*|[0-9]*\.?[0-9]+)\s*/g;
    this.tokens = expr.split(reg).filter(each => !each.match(/^\s*$/));
    this.idx = 0;
  }


  getToken() {
    if (this.idx > this.tokens.length) {
      throw new Error('Wrong Expression');
    }
    return this.tokens[this.idx++];
  }

  ungetToken() {
    this.idx--;
  }

  copy() {
    return this.tokens.slice(this.idx);
  }
}

class Interpreter {
  constructor() {
    this.vars = {};
    this.functions = {};
  }

  input(expr) {
    if (expr === '') return '';
    const tokens = new Token(expr);
    console.log(tokens);
    const token = tokens.getToken();
    if (token === 'fn') {
      this.parseFunction(tokens);
      return '';
    }
    tokens.ungetToken();
    return this.parseExpression(tokens);
  }

  parseFunction(tokens) {
    const fn = tokens.getToken();
    if (this.vars[fn] !== undefined) {
      throw new Error(`${fn} is a variable`);
    }

    const params = [];
    const paramObj = {};

    let token = tokens.getToken();
    while (token !== '=') {
      if (paramObj[token] !== undefined) {
        throw new Error('params conflict');
      }
      paramObj[token] = 1;
      params.push(token);
      token = tokens.getToken();
    }

    token = tokens.getToken();
    if (token !== '>') {
      throw new Error('illegal fucntion definition');
    }

    const block = tokens.copy();
    if (block.length === 0) {
      throw new Error('error');
    }

    for (let i = 0; i < block.length; i++) {
      token = block[i];
      if (/^[a-zA-Z]*$/.test(token)) {
        if (paramObj[token] === undefined && this.vars[token] === undefined) {
          throw new Error(`ERROR: Unknown identifier '${token}'`);
        }
        if (this.vars[token] !== undefined) {
          block[i] = this.vars[token];
        }
      }
    }
    this.functions[fn] = {
      params,
      block
    };

    return;
  }

  parseExpression(tokens, env = {}) {
    if (this.isAssignment(tokens)) {
      return this.parseAssignment(tokens);
    }

    let v1 = this.parseTerm(tokens, env);

    if (/^[a-zA-Z]*$/.test(v1)) {
      throw new Error(`undefined variable '${v1}'`);
    }

    for (;;) {
      const token = tokens.getToken();
      if (token !== '-' && token !== '+') {
        tokens.ungetToken();
        break;
      }

      const v2 = this.parseTerm(tokens, env);

      if (token === '+') {
        v1 += v2;
      } else if (token === '-') {
        v1 -= v2;
      } else {
        tokens.ungetToken();
      }
    }

    return v1;
  }

  isAssignment(tokens) {
    const _ = tokens.getToken();
    const token = tokens.getToken();
    tokens.ungetToken();
    tokens.ungetToken();
    return token === '=';
  }

  parseAssignment(tokens) {
    const name = tokens.getToken();
    tokens.getToken(); // pass `=`
    const v = this.parseExpression(tokens);
    this.vars[name] = v;
    return this.vars[name];
  }

  parseTerm(tokens, env = {}) {
    let v1 = this.parsePrimaryExpression(tokens, env);

    for (;;) {
      const token = tokens.getToken();
      if (token !== '*' && token !== '/' && token !== '%') {
        tokens.ungetToken();
        break;
      }

      const v2 = this.parsePrimaryExpression(tokens, env);

      if (token === '*') {
        v1 *= v2;
      } else if (token === '/') {
        v1 /= v2;
      } else if (token === '%') {
        v1 %= v2;
      }
    }

    return v1;
  }

  parsePrimaryExpression(tokens, env = {}) {
    let token = tokens.getToken();

    let minusFlag = false;
    if (token === '-') {
      minusFlag = true;
    } else {
      tokens.ungetToken();
    }

    token = tokens.getToken();

    let v;
    if (/\d/.test(token)) {
      v = Number(token);
      if (minusFlag) v = -v;
    } else if (/[a-zA-Z]/.test(token)) {
      console.log('exp', token, env);
      if (this.functions[token] !== undefined) {
        const fn = this.functions[token];
        v = this.executeFunction(fn, tokens);
      } else if (env[token] !== undefined) {
        v = this.parsePrimaryExpression(new Token(env[token]));
      } else if (this.vars[token] !== undefined) {
        v = this.vars[token];
      } else {
        v = token;
      }
    } else if (token === '(') {
      v = this.parseExpression(tokens, env);
      tokens.getToken(); // pass `)`
    } else {
      tokens.ungetToken();
    }

    return v;
  }

  executeFunction(fn, tokens) {
    const { params, block } = fn;
    const paramObj = {};
    for (const p of params) {
      const v = tokens.getToken();
      paramObj[p] = v;
    }
    console.log('ppp', paramObj);
    return this.parseExpression(new Token(block.join('')), paramObj);
  }
}

const readline = require('readline');

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

rl.setPrompt('>>> ');
rl.prompt();

const iter = new Interpreter();

rl.on('line', line => {
  const res = iter.input(line);
  console.log(res);
  rl.prompt();
});

rl.on('close', _ => {
  process.exit(0);
});

process.on('uncaughtException', e => {
  console.log(e.message);
  rl.prompt();
});

File: 1kyu/simple_interactive_interpreter.py
import re

class Token:
    def __init__(self, expr = ""):
        regex = re.compile("\s*(=>|[-+*\/\%=\(\)]|[A-Za-z_][A-Za-z0-9_]*|[0-9]*\.?[0-9]+)\s*")
        tokens = regex.findall(str(expr))
        self._token = [s for s in tokens if not s.isspace()]
        self._idx = 0
        
    def getToken(self):
        if self._idx >= len(self._token):
            return None
        token = self._token[self._idx]
        self._idx += 1
        return token

    def ungetToken(self):
        self._idx -= 1

    def copy(self):
        return self._token[self._idx:]
    
    def idx(self):
        return self._idx

    def len(self):
        return len(self._token) - self._idx

    
class Interpreter:
    def __init__(self):
        self.vars = {}
        self.functions = {}

    def input(self, expression):
        tokens = Token(expression)
        if tokens.len() == 0:
            return ''
        token = tokens.getToken()
        if token == 'fn':
            self.parseFunction(tokens)
            return ''
        tokens.ungetToken()
        v = self.parseExpression(tokens)
        if tokens.len() != 0:
            raise Exception('Error')
        return v

    def parseFunction(self, tokens):
        fn = tokens.getToken()
        if fn in self.vars:
            raise Exception('Error is val')

        params = []
        paramObj = {}

        token = tokens.getToken()
        while token != '=>':
            if token in paramObj:
                raise Exception('Error conflict para')
            paramObj[token] = 1
            params.append(token)
            token = tokens.getToken()

        block = tokens.copy()

        if len(block) == 0:
            raise Exception('Error')

        for i in xrange(0, len(block)):
            token = block[i]
            if self.isIdentifier(token) and token not in paramObj:
                raise Exception('Error')

        self.functions[fn] = {
            "params": params,
            "block": block
        }

    def parseExpression(self, tokens, env = {}):
        if self.isAssignment(tokens):
            return self.parseAssignment(tokens)

        v1 = self.parseTerm(tokens, env)
        if v1 is None:
            return v1
        if self.isIdentifier(v1):
            raise Exception('Error: Undefined variable')

        while True:
            token = tokens.getToken()
            if token is None:
                return v1
            if token != '-' and token != '+':
                tokens.ungetToken()
                break
            v2 = self.parseTerm(tokens, env)
            if v2 is None:
                return v1
            if token == '+':
                v1 += v2
            elif token == '-':
                v1 -= v2
            else:
                tokens.ungetToken()
        return v1

    def isAssignment(self, tokens):
        if tokens.len() <= 2:
            return False
        tokens.getToken()
        token = tokens.getToken()
        tokens.ungetToken()
        tokens.ungetToken()
        return token == '=';

    def parseAssignment(self, tokens):
        name = tokens.getToken()
        if name in self.functions:
            raise Exception('Error duplicate fn')
        token = tokens.getToken()
        if token is None:
            return None
        v = self.parseExpression(tokens);
        if v is None:
            return v
        self.vars[name] = v
        return v

    def parseTerm(self, tokens, env = {}):
        v1 = self.parsePrimaryExpression(tokens, env)
        if v1 is None:
            return v1
        while True:
            token = tokens.getToken()
            if token is None:
                return v1
            if token not in '*/%':
                tokens.ungetToken()
                break

            v2 = self.parsePrimaryExpression(tokens, env)
            if v2 is None:
                return v1
            if token == '*':
                v1 *= v2
            elif token == '/':
                v1 //= v2
            elif token == '%':
                v1 %= v2
            else:
                pass #tokens.ungetToken()
        return v1

    def parsePrimaryExpression(self, tokens, env = {}):
        token = tokens.getToken()
        if token is None:
            return None
        minusFlag = False
        if token == '-':
            minusFlag = True
        else:
            tokens.ungetToken()

        token = tokens.getToken()
        if token is None:
            return None
        v = None
        if token.isdigit():
            v = int(token)
            if minusFlag:
                v = -v
        elif self.isIdentifier(token):
            if token in self.functions:
                fn = self.functions[token]
                v = self.executeFunction(fn, tokens)
            elif token in env:
                v = self.parsePrimaryExpression(Token(env[token]))
            elif token in self.vars:
                v = self.vars[token]
            else:
                v = token
        elif token == '(':
            v = self.parseExpression(tokens, env)
            tokens.getToken() # pass ')'
        else:
            tokens.ungetToken()
        return v

    def executeFunction(self, fn, tokens):
        params = fn["params"]
        block = fn["block"]
        paramObj = {}
        for p in params:
            paramObj[p] = self.parseExpression(tokens)
        return self.parseExpression(Token(''.join(block)), paramObj)

    def isIdentifier(self, token):
        return re.match(r'^[a-zA-Z]*$', str(token))
